main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -0,0 +1,214 @@
+ * 2025 Jakob Schneider Villumsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.optimizer
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.Occur.*
+import ca.uwaterloo.flix.language.ast.OccurrenceAst.{DefContext, Expr, Occur}
+import ca.uwaterloo.flix.language.ast.{AtomicOp, OccurrenceAst, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.util.collection.CofiniteSet
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+
+import java.util.concurrent.ConcurrentLinkedQueue
+
+/**
+  * The inliner optionally performs beta-reduction at call-sites.
+  */
+object Inliner {
+
+  /**
+    * Performs inlining on the given AST `root`.
+    */
+  def run(root: OccurrenceAst.Root)(implicit flix: Flix): OccurrenceAst.Root = {
+    implicit val sctx: SharedContext = SharedContext.mk()
+    val defs = ParOps.parMapValues(root.defs)(visitDef(_)(root, sctx, flix))
+    val newRoot = OccurrenceAst.Root(defs, root.enums, root.structs, root.effects, root.mainEntryPoint, root.entryPoints, root.sources)
+    newRoot
+  }
+
+  /**
+    * Performs expression inlining on the given definition `def0`.
+    * Converts definition from [[OccurrenceAst]] to [[OccurrenceAst]].
+    */
+  private def visitDef(def0: OccurrenceAst.Def)(implicit root: OccurrenceAst.Root, sctx: SharedContext, flix: Flix): OccurrenceAst.Def = def0 match {
+    case OccurrenceAst.Def(sym, fparams, spec, exp, ctx, loc) =>
+      val e = visitExp(exp, Context.Empty)(sym, root, sctx, flix)
+      OccurrenceAst.Def(sym, fparams, spec, e, ctx, loc)
+  }
+
+  /**
+    * Performs inlining operations on the expression `exp0` from [[Expr]].
+    * Returns a [[Expr]]
+    */
+  private def visitExp(exp0: Expr, ctx0: Context)(implicit sym0: Symbol.DefnSym, root: OccurrenceAst.Root, sctx: SharedContext, flix: Flix): Expr = exp0
+
+  /**
+    * Returns `true` if `def0` should be inlined.
+    */
+  private def canInlineDef(defCtx: DefContext, ctx: Context): Boolean = {
+    val mayInline = !defCtx.isSelfRecursive && !ctx.currentlyInlining
+    val shouldInline = defCtx.isDirectCall
+    mayInline && shouldInline
+  }
+
+  private def isPure(eff0: Type): Boolean = {
+    eff0 == Type.Pure
+  }
+
+  /**
+    * Checks if `occur` is Dead.
+    */
+  private def isDead(occur: OccurrenceAst.Occur): Boolean = occur match {
```

Comment:
Occur is very small.

Lets make all the pattern matches on it exhaustive. That way, if we ever extend it, we will get the compiler errors we want to help us.